examples/custom_search_method/core_search_runner/asha.py

```diff
@@ -0,0 +1,483 @@
+    # Invoked only once, when starting a new experiment. Creates initial list
+    # of operations.
+    # In this example, we create and submit operations for first N trials, such that:
+    #   1) each trial is assigned a unique request_id and every operation
+    #      contains request_id of a trial it refers to;
+    #   2) each trial is initialized with two operations:
+    #      -> "Create" operation that takes in trial's request_id and hyperparameters;
+    #         in this example hyperparamters are generated by user-defined method
+    #         sample_params(),
+    #      -> "ValidateAfter" operation that takes in trial's request_id and number of
+    #         units (batches or epochs) that the model is trained for before validation;
+    #         units selection is made in the custom_config.yaml.
+    #
+    # Note: the order in which trials are created is not guaranteed.
+    def initial_operations(self) -> List[Operation]:
+        ops: List[Operation] = []
+        N = self._get_max_concurrent_trials()
+
+        for _ in range(0, N):
+            create = Create(
+                request_id=uuid.uuid4(),
+                hparams=self.sample_params(),
+                checkpoint=None,
+            )
+            ops.append(create)
+            ops.append(
+                ValidateAfter(
+                    request_id=create.request_id,
+                    length=self.asha_search_state.rungs[0].units_needed,
+                )
+            )
+
+            self.asha_search_state.trial_rungs[create.request_id] = 0
+            self.asha_search_state.pending_trials += 1
+
+        return ops
+
+    ############################################################################
+    # Invoked when a trial with specific request_id is created.
+    # In this example, ASHASearchMethodState is updated with
+    # information about a trial's progress and no new operations are created.
+    def on_trial_created(self, request_id: uuid.UUID) -> List[Operation]:
+        self.asha_search_state.rungs[0].outstanding_trials += 1
+        self.asha_search_state.trial_rungs[request_id] = 0
+        return []
+
+    ############################################################################
+    # Invoked when a validation for a trial with specific request_id is completed.
+    # Provides a result of validation according to the metric specified in the
+    # custom_config.yaml: validation_loss.
+    # In this example,ASHASearchMethodState statistics are updated,
+    # and based on the current number of trials and their metric values,
+    # ASHA decides whether to:
+    #    (1) promote a trial, i.e., to continue training for the trial
+    #    (2) start a new trial,
+    #    (3) close experiment if all trials are completed and maximum number of
+    #        trials is reached.
+    def on_validation_completed(self, request_id: uuid.UUID, metric: float) -> List[Operation]:
```

Comment:
if we wind up passing the op training length here to `on_validation_completed`, is there a way we can show using that somehow for asha? perhaps it'd be useful as an arg to `_promote_async`